src/compiler/checker.ts

```diff
@@ -7608,11 +7608,18 @@ namespace ts {
                     // itself. When inferring from a type to itself we effectively find all type parameter
                     // occurrences within that type and infer themselves as their type arguments.
                     let matchingTypes: Type[];
-                    for (const t of (<UnionOrIntersectionType>target).types) {
-                        if (typeIdenticalToSomeType(t, (<UnionOrIntersectionType>source).types)) {
+                    for (const t of (<UnionOrIntersectionType>source).types) {
+                        if (typeIdenticalToSomeType(t, (<UnionOrIntersectionType>target).types)) {
                             (matchingTypes || (matchingTypes = [])).push(t);
                             inferFromTypes(t, t);
                         }
+                        else if (t.flags & (TypeFlags.NumberLiteral | TypeFlags.StringLiteral)) {

```

Comment:
Yes. The `boolean` type is just another name for the union `true | false`, so when `boolean` is included in a union type it is the same as including `true` and `false`. Therefore we need to special handling for the boolean literals. Similar situation for enum types. However, the `number` and `string` types effectively have infinite sets of literal values we can't flatten them and must handle them specially.